doc/proposals/scorecard-custom-tests.md

```diff
@@ -0,0 +1,403 @@
+
+Today there are 2 internal scorecard tests that make use of the scorecard proxy, with these tests migrating to the custom test format, the scorecard proxy is likely not necessary which would also reduce the overall scorecard implementation complexity.
+
+#### Building the Proposed Design
+
+The proposed design was influenced greatly by the KUDO testing framework in various concepts, however, the proposed design doesn’t necessarily have to include KUDO as a dependency.  A decision of whether or not to incorporate KUDO testing APIs within the Red Hat solution needs to be reached.  Some pro’s and con’s to using KUDO testing frameworks are listed below.  If KUDO’s testing framework is not incorporated, then a custom built set of equivalent capabilities would need to be developed.
+The advantage of using Kuttl as the test framework in scorecard would include:
+
+ * Pre-built, works, community, evolving.
+ * Ready to work with, code exists and mostly works, a starting point exists.
+ * Handles the basic test case, that being to test Kube resource state
+ * Written in golang, making integration into scorecard a possibility.
+
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+### Upgrade / Downgrade Strategy
+
+To implement this custom test functionality, additions or changes to the existing scorecard configuration file and command line flags would be necessary and would not be supported in previous scorecard releases.
+
+## Implementation History
+
+ * Proposal initiated in March 2020.
+
+## Drawbacks
+
+Drawbacks of using Kuttl within this proposed design might include:
+
+ * Immaturity of Kuttl and KUDO in general, integration into Operator SDK will require Redhat to possibly contribute upstream to the Kuttl project itself in order to meet the needs of the Operator SDK.
+  * <https://github.com/kudobuilder/kuttl/issues/16#issuecomment-589242819>
+  * <https://github.com/kudobuilder/kuttl/issues/12>
+  * <https://github.com/kudobuilder/kudo/blob/master/keps/0008-operator-testing.md>
+  * <https://github.com/kudobuilder/kuttl/pull/5>
+
+ * Dependency on a community project for enhancements or stability of APIs
+ * Doesn’t do much in terms of complex testing
+ * Needs some kind of regex capability for a more robust YAML testing capability
+ * Will require integration work to include it into the scorecard code.
+ * KUDO doesn’t currently make it easy to get the test Pod log output, this is useful to determine why a test might have failed, without it being built-into Kudo, the scorecard would need to read the Pod logs to present back to the end user.
+
+## Alternatives
+
+The alternatives to this design proposal might include:
+ * A custom built test framework be designed and developed solely for the purpose of implementing Operator SDK scorecard custom tests.
+ * Other existing/open-source test frameworks could be evaluated for use within the SDK for implementing the desired custom test functionality.
+
```

Comment:
Hi @jmccormick2001, 
 
Also, could not be an alternative: 

- we create an operator which will have an API such as kutll as:(CR/CRD)
- in this CR the user will do all setup 
- then, we will in the reconcile of this operator we will compare the expected results (assert files) with the objects created in the cluster.  

Also, to use it we can have a command such as `operator-sdk assertcard` that will deploy the user operator + our assertcard one which will do the checks. Then, all logs will be in the container as it is for all operators and configured as we wish.   

See that it would give to us a lot of flexibility and since would be built with an SDK would be 100% supportable and easy to keep maintained/updated as producing a matrix of compatibility. Also, it would allow us to add all OCP schemas as any third-party schemas which can be informed by the users via its APIs as well. 

Is it make sense? What do you think?

Reply:
sounds possible, but a different model for sure than what this proposal offers.  This proposed design has the test tool being standalone, not deployed to the Kube cluster, but merely a cluster client.  The issue I see with your suggestion is the deploying and maintenance of another operator that end users would need to deploy in order to run tests.  Possible and worth some discussion however.

Reply:
The same solution as we spoke about can be done to run as standalone as `operator-sdk run---local` and `operator-sdk test` does.